File: tests/parity_check.py
"""Generate a Python physics fixture and compare against the TypeScript port.

Usage:
  python tests/parity_check.py              # dump fixture + run TS check via npx tsx
  python tests/parity_check.py --dump-only  # only write fixture JSON
"""

from __future__ import annotations

import argparse
import json
import os
import subprocess
import sys

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
sys.path.insert(0, ROOT)

from environment import FishingMinigameEnv
from portable_rng import PortableRNG

FIXTURE_PATH = os.path.join(ROOT, "competition-website", "scripts", "parity-fixture.json")
TS_CHECK = os.path.join(ROOT, "competition-website", "scripts", "parity-check.ts")


def dump_fixture(seed: int = 12345, fish_name: str = "Carp", steps: int = 200) -> dict:
    env = FishingMinigameEnv(
        render_mode=None, seed=seed, fish_name=fish_name, augment_fish=False
    )
    env.fish_name = fish_name
    obs = env.reset(seed=seed)

    # Deterministic action sequence (press pattern) — not random, so RNG stays in physics
    actions = [1 if (i % 4) < 2 else 0 for i in range(steps)]

    frames = [
        {
            "t": 0,
            "bobberPosition": env.bobberPosition,
            "bobberSpeed": env.bobberSpeed,
            "bobberTargetPosition": env.bobberTargetPosition,
            "bobberBarPos": env.bobberBarPos,
            "bobberBarSpeed": env.bobberBarSpeed,
            "floaterSinkerAcceleration": env.floaterSinkerAcceleration,
            "distanceFromCatching": env.distanceFromCatching,
            "bobberInBar": bool(env.bobberInBar),
            "obs": obs.tolist(),
            "reward": None,
        }
    ]

    for i, action in enumerate(actions):
        obs, reward, done, _ = env.step(action)
        frames.append(
            {
                "t": i + 1,
                "action": action,
                "bobberPosition": env.bobberPosition,
                "bobberSpeed": env.bobberSpeed,
                "bobberTargetPosition": env.bobberTargetPosition,
                "bobberBarPos": env.bobberBarPos,
                "bobberBarSpeed": env.bobberBarSpeed,
                "floaterSinkerAcceleration": env.floaterSinkerAcceleration,
                "distanceFromCatching": env.distanceFromCatching,
                "bobberInBar": bool(env.bobberInBar),
                "obs": obs.tolist(),
                "reward": float(reward),
                "done": bool(done),
            }
        )
        if done:
            break

    fixture = {
        "seed": seed,
        "fish": {
            "name": env.current_fish["name"],
            "difficulty": int(env.current_fish["difficulty"]),
            "behaviour": env.current_fish["behaviour"],
        },
        "actions": actions[: len(frames) - 1],
        "frames": frames,
    }
    env.close()
    return fixture


def check_rng_selftest() -> None:
    """Sanity: PortableRNG produces a stable known sequence."""
    rng = PortableRNG(42)
    vals = [rng.next_uint32() for _ in range(5)]
    # Locked golden values — update TS test if these change
    expected = [2581720956, 1925393290, 3661312704, 2876485805, 750819978]
    if vals != expected:
        raise AssertionError(f"PortableRNG golden mismatch: {vals} != {expected}")
    print("PortableRNG self-test OK")


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--dump-only", action="store_true")
    parser.add_argument("--seed", type=int, default=12345)
    parser.add_argument("--fish", type=str, default="Carp")
    parser.add_argument("--steps", type=int, default=200)
    args = parser.parse_args()

    os.chdir(ROOT)
    check_rng_selftest()

    fixture = dump_fixture(seed=args.seed, fish_name=args.fish, steps=args.steps)
    os.makedirs(os.path.dirname(FIXTURE_PATH), exist_ok=True)
    with open(FIXTURE_PATH, "w") as f:
        json.dump(fixture, f)
    print(f"Wrote fixture: {FIXTURE_PATH} ({len(fixture['frames'])} frames)")

    if args.dump_only:
        return 0

    # Run TypeScript parity check
    npx = "npx.cmd" if sys.platform == "win32" else "npx"
    cmd = [npx, "--yes", "tsx", TS_CHECK, FIXTURE_PATH]
    print("Running:", " ".join(cmd))
    result = subprocess.run(cmd, cwd=os.path.join(ROOT, "competition-website"), shell=False)
    return result.returncode


if __name__ == "__main__":
    sys.exit(main())

File: competition-website/scripts/parity-check.ts
/**
 * Compare TypeScript FishingGame against a Python-generated fixture.
 * Run via: npx tsx scripts/parity-check.ts [fixture.json]
 */
import * as fs from 'fs';
import * as path from 'path';
import { FishingGame, Fish } from '../lib/game-logic';
import { PortableRNG } from '../lib/portable-rng';

const TOL = 1e-9;

function assertClose(label: string, a: number, b: number, tol = TOL) {
    if (Math.abs(a - b) > tol) {
        throw new Error(`${label}: TS=${a} Python=${b} (Δ=${Math.abs(a - b)})`);
    }
}

function rngSelfTest() {
    const rng = new PortableRNG(42);
    const vals = Array.from({ length: 5 }, () => rng.nextUint32());
    const expected = [2581720956, 1925393290, 3661312704, 2876485805, 750819978];
    for (let i = 0; i < expected.length; i++) {
        if (vals[i] !== expected[i]) {
            throw new Error(`PortableRNG golden mismatch at ${i}: ${vals[i]} !== ${expected[i]}`);
        }
    }
    console.log('PortableRNG self-test OK');
}

function main() {
    rngSelfTest();

    const fixturePath =
        process.argv[2] || path.join(__dirname, 'parity-fixture.json');
    const fixture = JSON.parse(fs.readFileSync(fixturePath, 'utf8'));

    const fish: Fish = fixture.fish;
    const game = new FishingGame(fish, fixture.seed, { augmentFish: false });
    game.reset(fixture.seed);

    const f0 = fixture.frames[0];
    assertClose('t0.bobberPosition', game.state.bobberPosition, f0.bobberPosition);
    assertClose('t0.bobberBarPos', game.state.bobberBarPos, f0.bobberBarPos);
    for (let i = 0; i < f0.obs.length; i++) {
        assertClose(`t0.obs[${i}]`, game.getObservation()[i], f0.obs[i]);
    }

    let mismatches = 0;
    for (let i = 0; i < fixture.actions.length; i++) {
        const action = fixture.actions[i] as 0 | 1;
        const result = game.step(action);
        const expected = fixture.frames[i + 1];

        const checks: [string, number, number][] = [
            ['bobberPosition', game.state.bobberPosition, expected.bobberPosition],
            ['bobberSpeed', game.state.bobberSpeed, expected.bobberSpeed],
            ['bobberTargetPosition', game.state.bobberTargetPosition, expected.bobberTargetPosition],
            ['bobberBarPos', game.state.bobberBarPos, expected.bobberBarPos],
            ['bobberBarSpeed', game.state.bobberBarSpeed, expected.bobberBarSpeed],
            [
                'floaterSinkerAcceleration',
                game.state.floaterSinkerAcceleration,
                expected.floaterSinkerAcceleration,
            ],
            ['distanceFromCatching', game.state.distanceFromCatching, expected.distanceFromCatching],
            ['reward', result.reward, expected.reward],
        ];

        for (const [label, a, b] of checks) {
            try {
                assertClose(`t${i + 1}.${label}`, a, b);
            } catch (e) {
                console.error(String(e));
                mismatches++;
            }
        }

        if (Boolean(game.state.bobberInBar) !== Boolean(expected.bobberInBar)) {
            console.error(
                `t${i + 1}.bobberInBar: TS=${game.state.bobberInBar} Python=${expected.bobberInBar}`
            );
            mismatches++;
        }

        for (let j = 0; j < expected.obs.length; j++) {
            try {
                assertClose(`t${i + 1}.obs[${j}]`, result.obs[j], expected.obs[j]);
            } catch (e) {
                console.error(String(e));
                mismatches++;
            }
        }

        if (mismatches > 20) {
            console.error('Too many mismatches — aborting');
            process.exit(1);
        }

        if (result.done) break;
    }

    if (mismatches > 0) {
        console.error(`FAIL: ${mismatches} mismatches`);
        process.exit(1);
    }
    console.log(`PASS: ${fixture.actions.length} steps match Python fixture`);
}

main();
